## Constants --------------------------------------------------------------
# NCBI server (README.txt - https://ftp.ncbi.nlm.nih.gov/genomes/all/README.txt)
NCBI_FTP_HOST = 'ftp.ncbi.nlm.nih.gov'
NCBI_FTP_URL = 'https://ftp.ncbi.nlm.nih.gov/genomes/all'

# ENSEMBL Rapid Release server
# ENSEMBL_FTP_HOST = 'ftp.ensembl.org'
# ENSEMBL_FTP_URL = 'http://ftp.ensembl.org/pub/'
# ENSEMBL_RAPID_FTP_URL = ENSEMBL_FTP_URL + 'rapid-release/'
# ENSEMBL_RM_FTP_URL = 'https://ftp.ebi.ac.uk/pub/databases/ensembl/repeats/unfiltered_repeatmodeler/species/'

# ENSEMBL Beta server
ENSEMBL_BETA_FTP_URL = 'http://ftp.ebi.ac.uk/pub/ensemblorganisms/'
ENSEMBL_TREE_JSON = 'https://raw.githubusercontent.com/snu-cdrc/gencube/refs/heads/main/data-aggregator/ensembl-beta_ftp_structure.json'

# UCSC GenArk server (the Earth BioGenome Project, the Vertebrate Genomes Project, the Telomere-to-Telomere Consortium, and other related projects.)
GENARK_URL = 'https://hgdownload.soe.ucsc.edu/hubs/'

# Zoonomia TOGA (200 mammals project)
ZOONOMIA_URL = 'https://genome.senckenberg.de/download/TOGA/'
ZOONOMIA_META = "https://github.com/snu-cdrc/gencube/raw/refs/heads/main/data-aggregator/zoonomia_meta.csv"

# UCSC-Kent applications
UCSC_KENT_URL = 'https://hgdownload.soe.ucsc.edu/admin/exe'

## Lists --------------------------------------------------------------
# SRA advanced search index lists
SRA_SEARCH_INDEX_JSON = 'https://raw.githubusercontent.com/snu-cdrc/gencube/refs/heads/main/data-aggregator/sra_advanced_index.json'

# NCBI Assembly Report
LS_NCBI_ASSEMBLY_META_KEY = [
    'Synonym_GCA',
    'Synonym_GCF',
    'LastMajorReleaseAccession',
    'LatestAccession',
    'AssemblyName',
    'UCSCName',
    'EnsemblName',
    'Taxid',
    'Organism',
    'SpeciesTaxid',
    'SpeciesName',
    'AssemblyType',
    'AssemblyStatus',
    'WGS',
    'GB_BioProjects',
    'GB_Projects',
    'RS_BioProjects',
    'RS_Projects',
    'BioSampleAccn',
    'Biosource',
    'Coverage',
    'PartialGenomeRepresentation',
    'Primary',
    'AssemblyDescription',
    'ReleaseLevel',
    'ReleaseType',
    'AsmReleaseDate_GenBank',
    'AsmReleaseDate_RefSeq',
    'SeqReleaseDate',
    'AsmUpdateDate',
    'SubmissionDate',
    'LastUpdateDate',
    'SubmitterOrganization',
    'RefSeq_category',
    'AnomalousList',
    'ExclFromRefSeq',
    'PropertyList',
    'FromType',
    'ContigN50',
    'ScaffoldN50',
    'Busco',
    'SortOrder',
    'Meta',
    ]
LS_NCBI_ASSEMBLY_META_LABEL = [
    'GenBank',
    'RefSeq',
    'Last major release accession',
    'Latest accession',
    'Assembly name',
    'UCSC',
    'Ensembl name',
    'Taxid',
    'Organism',
    'Species taxid',
    'Species name',
    'Assembly type',
    'Level',
    'WGS',
    'GB_BioProjects',
    'GB_Projects',
    'RS_BioProjects',
    'RS_Projects',
    'BioSampleAccn',
    'Biosource',
    'Coverage',
    'PartialGenomeRepresentation',
    'Primary',
    'AssemblyDescription',
    'ReleaseLevel',
    'ReleaseType',
    'Release', # GenBank
    'Release (RefSeq)',
    'SeqReleaseDate',
    'Update',
    'SubmissionDate',
    'LastUpdateDate',
    'Organization',
    'RefSeq_category',
    'AnomalousList',
    'ExclFromRefSeq',
    'PropertyList',
    'FromType',
    'ContigN50',
    'ScaffoldN50',
    'Busco',
    'SortOrder',
    'Meta',
    ]
LS_GENCUBE_GENOME_LABEL = [
    'Assembly name',
    'Taxid',
    'Release',
    'Level',
    'NCBI',
    'UCSC',
    'GenArk',
    'Ensembl',
    ]
LS_GENCUBE_GENSET_LABEL = [
    'Assembly name',
    'Taxid',
    'Release',
    'Level',
    'NCBI',
    'UCSC',
    'GenArk',
    'Ensembl',
    'Zoonomia',
    ]
LS_GENCUBE_SEQUENCE_LABEL = [
    'Assembly name',
    'Taxid',
    'Release',
    'Level',
    'NCBI',
    'UCSC',
    'Ensembl',
    ]
LS_GENCUBE_ANNOTATION_LABEL = [
    'Assembly name',
    'Taxid',
    'Release',
    'Level',
    'NCBI',
    'UCSC',
    'GenArk',
    ]
LS_GENCUBE_CROSSGENOME_LABEL = [
    'Assembly name',
    'Taxid',
    'Release',
    'Level',
    'NCBI',
    'UCSC',
    'Ensembl',
    'Zoonomia',
    ]

LS_ASSEMBLY_REPORT_LABEL = [
    'Sequence-Name',
    'Sequence-Role',
    'Assigned-Molecule',
    'Assigned-Molecule-Location/Type',
    'GenBank-Accn',
    'Relationship',
    'RefSeq-Accn',
    'Assembly-Unit',
    'Sequence-Length',
    'UCSC-style-name'
    ]

LS_SRA_META_STUDY_KEY = [
    'EXPERIMENT.STUDY_REF.@accession',
    'STUDY.@alias',
    'SAMPLE.SAMPLE_LINKS.SAMPLE_LINK.XREF_LINK.LABEL',
    'EXPERIMENT.STUDY_REF.IDENTIFIERS.EXTERNAL_ID.#text',
    'STUDY.IDENTIFIERS.EXTERNAL_ID.#text',
    'STUDY.DESCRIPTOR.STUDY_TITLE',
    'STUDY.DESCRIPTOR.STUDY_ABSTRACT',
    'EXPERIMENT.DESIGN.LIBRARY_DESCRIPTOR.LIBRARY_CONSTRUCTION_PROTOCOL',
    'SUBMISSION.@center_name',
    'Organization.Contact.Address.Country',
    'Organization.Address.Country',
    'RUN_SET.RUN.@published',
    ]
LS_SRA_META_STUDY_LABEL = [
    'Study',
    'GSE',
    'BioProject_alt1',
    'BioProject_alt2',
    'BioProject_alt3',
    'Title',
    'Abstract',
    'Protocol',
    'Submission',
    'Country_alt1',
    'Country_alt2',
    'Published',
    ]
LS_SRA_META_EXPERIMENT_KEY = [
    'EXPERIMENT.@accession',
    'SAMPLE.@accession',
    'SAMPLE.@alias',
    'SAMPLE.SAMPLE_NAME.SCIENTIFIC_NAME',
    'SAMPLE.SAMPLE_NAME.TAXON_ID',
    'EXPERIMENT.DESIGN.LIBRARY_DESCRIPTOR.LIBRARY_NAME',
    'EXPERIMENT.DESIGN.LIBRARY_DESCRIPTOR.LIBRARY_STRATEGY',
    'EXPERIMENT.DESIGN.LIBRARY_DESCRIPTOR.LIBRARY_SOURCE',
    'EXPERIMENT.PLATFORM.ILLUMINA.INSTRUMENT_MODEL',
    'EXPERIMENT.DESIGN.LIBRARY_DESCRIPTOR.LIBRARY_SELECTION',
    'EXPERIMENT.TITLE',
    'Pool.Member.@sample_title', # Alternative: 'SAMPLE.TITLE',
    'EXPERIMENT.DESIGN.DESIGN_DESCRIPTION',
    'SAMPLE.SAMPLE_ATTRIBUTES.SAMPLE_ATTRIBUTE',
    'EXPERIMENT.EXPERIMENT_ATTRIBUTES.EXPERIMENT_ATTRIBUTE',
    'RUN_SET.RUN.SRAFiles.SRAFile',
    ]
LS_SRA_META_EXPERIMENT_LABEL = [
    'Experiment',
    'Sample',
    'GSM',
    'Scientific name',
    'Taxon id',
    'Library name',
    'Strategy',
    'Source',
    'Platform',
    'Selection',
    'Experiment title',
    'Sample title',
    'Design description',
    'Sample attribute',
    'Experiment attribute',
    'File information',
    ]

## Dictionaries --------------------------------------------------------------
DIC_ZOONOMIA = {
    'human': 'human_hg38_reference',
    'mouse': 'mouse_mm10_reference',
    'chicken': 'chicken_galGal6_reference',
    'greenSeaturtle': 'greenSeaturtle_HLcheMyd2_reference',
    'pikePerch': 'pikePerch_HLsanLuc1_reference',
    'purpleSeaUrchin': 'purpleSeaUrchin_HLstrPur5_reference',
    'redEaredSlideTurtle': 'redEaredSlideTurtle_HLtraScrEle1_reference',
    'thaleCress': 'thaleCress_HLParaTha1_reference',
    'tobaccoHawkmoth': 'tobaccoHawkmoth_HLmanSex2_reference'
    }

"""
LS_STRATEGY = [
    # Genomic
    'wgs',
    'wga',
    'wxs',
    'targeted_capture',
    'synthetic_long_read',
    'gbs',
    'rad_seq',
    'tn_seq',
    'clone_end',
    # Genomic or Transcriptomic,
    'amplicon',
    'clone',
    # Transcriptomic
    'rna_seq',
    'mrna_seq',
    'ncrna_seq',
    'ribo_seq',
    'rip_seq',
    'mirna_seq',
    'ssrna_seq',
    'est',
    'fl_cdna',
    # Epigenomic - Chromatin Accessibility
    'atac_seq',
    'dnase_hypersensitivity',
    'faire_seq',
    # Epigenomic - DNA-Protein Binding
    'chip_seq',
    'chip',
    # Epigenomic - Methylome
    'mre_seq',
    'bisulfite_seq',
    'mbd_seq',
    'medip_seq',
    # Epigenomic - Chromatin Interaction and Structural Analysis
    'hi_c',
    'chia_pet',
    'tethered_chromatin_conformation_capture',
    ]
LS_SOURCE = [
    'genomic',
    'genomic_single_cell',
    'transcriptomic',
    'transcriptomic_single_cell',
    'metagenomic',
    'metatranscriptomic',
    'synthetic',
    'viral_rna',
    'other',
    ]
LS_PLATFORM = [
    'abi_solid',
    'bgiseq',
    'capillary',
    'complete_genomics',
    'dnbseq',
    'element',
    'genapsys',
    'genemind',
    'helicos',
    'illumina',
    'ion_torrent',
    'ls454',
    'oxford_nanopore',
    'pacbio_smrt',
    'tapestri',
    'ultima',
    'vela_diagnostics ',
    ]
LS_SELECTION = [
    '5_methylcytidine_antibody',
    'cage',
    'cdna',
    'cdna_oligo_dt',
    'cdna_randompriming',
    'chip',
    'chip_seq',
    'dnase',
    'hmpr',
    'hybrid_selection',
    'inverse_rrna',
    'mbd2_protein_methyl_cpg_binding_domain',
    'mda',
    'mf',
    'mnase',
    'msll',
    'oligo_dt',
    'other',
    'padlock_probes_capture_method',
    'pcr',
    'polya',
    'race',
    'random',
    'random_pcr',
    'reduced_representation',
    'repeat_fractionation',
    'restriction_digest',
    'rt_pcr',
    'size_fractionation',
    'unspecified',
    ]
LS_FILTER = [
    'sra_all',
    'sra_assembly',
    'sra_bioproject',
    'sra_bioproject_all',
    'sra_biosample',
    'sra_biosample_all',
    'sra_gap',
    'sra_gap_all',
    'sra_gds',
    'sra_genome',
    'sra_nuccore',
    'sra_nuccore_alignment',
    'sra_nuccore_wgs',
    'sra_omim',
    'sra_pmc',
    'sra_public',
    'sra_pubmed',
    'sra_taxonomy',
]
LS_PROPERTIES = [
    'aligned_data',
    'cloud_gs',
    'cloud_s3',
    'location_gs_us',
    'location_s3_us_east1',
    'location_s3_us_east_1',
    'location_s3_us_west_2',
    'filetype_10x_genomics_bam_file',
    'filetype_ab1',
    'filetype_activ_sars2_vcf',
    'filetype_archive',
    'filetype_archive/gzip',
    'filetype_assembled_contigs',
    'filetype_assembly/realign_summary',
    'filetype_assembly_of_unidentified_reads',
    'filetype_bai',
    'filetype_bam',
    'filetype_bam_header',
    'filetype_basemodification',
    'filetype_complete_genomics',
    'filetype_crai',
    'filetype_cram',
    'filetype_fast5',
    'filetype_fasta',
    'filetype_fastq',
    'filetype_geo_feature_count',
    'filetype_helicos',
    'filetype_illumina_native',
    'filetype_nanopore',
    'filetype_pacbio_base_modification_report',
    'filetype_pacbio_metadata',
    'filetype_pacbio_native',
    'filetype_realign_to_de_novo_assembly',
    'filetype_reference_fasta',
    'filetype_run',
    'filetype_run_realign',
    'filetype_run_zq',
    'filetype_sff',
    'filetype_solid_native',
    'filetype_source',
    'filetype_sra_lite',
    'filetype_sra_normalized',
    'filetype_srf',
    'filetype_tar_archive_of_complete_genomics_tree',
    'filetype_tenx',
    'filetype_vcf',
    'filetype_vcf_index',
    'filetype_vdbcache',
    'filetype_vdbcache_zq',
    'filetype_wgmlst_sig',
    'filetype_wgmlst_signature',
    'has_data',
    'instrument_454_gs',
    'instrument_454_gs_20',
    'instrument_454_gs_flx',
    'instrument_454_gs_flx_titanium',
    'instrument_454_gs_junior',
    'instrument_ab_310_genetic_analyzer',
    'instrument_ab_3130_genetic_analyzer',
    'instrument_ab_3130xl_genetic_analyzer',
    'instrument_ab_3500_genetic_analyzer',
    'instrument_ab_3500xl_genetic_analyzer',
    'instrument_ab_3730_genetic_analyzer',
    'instrument_ab_3730xl_genetic_analyzer',
    'instrument_ab_5500_genetic_analyzer',
    'instrument_ab_5500xl_genetic_analyzer',
    'instrument_ab_5500xl_w_genetic_analysis_system',
    'instrument_ab_solid_3_plus_system',
    'instrument_ab_solid_4_system',
    'instrument_ab_solid_4hq_system',
    'instrument_ab_solid_pi_system',
    'instrument_ab_solid_system',
    'instrument_ab_solid_system_2_0',
    'instrument_ab_solid_system_3_0',
    'instrument_bgiseq_50',
    'instrument_bgiseq_500',
    'instrument_complete_genomics',
    'instrument_dnbseq_g400',
    'instrument_dnbseq_g400_fast',
    'instrument_dnbseq_g50',
    'instrument_dnbseq_t7',
    'instrument_element_aviti',
    'instrument_fastaseq_300',
    'instrument_genexus',
    'instrument_genolab_m',
    'instrument_gridion',
    'instrument_gs111',
    'instrument_helicos_heliscope',
    'instrument_hiseq_x_five',
    'instrument_hiseq_x_ten',
    'instrument_illumina_genome_analyzer',
    'instrument_illumina_genome_analyzer_ii',
    'instrument_illumina_genome_analyzer_iix',
    'instrument_illumina_hiscansq',
    'instrument_illumina_hiseq_1000',
    'instrument_illumina_hiseq_1500',
    'instrument_illumina_hiseq_2000',
    'instrument_illumina_hiseq_2500',
    'instrument_illumina_hiseq_3000',
    'instrument_illumina_hiseq_4000',
    'instrument_illumina_hiseq_x',
    'instrument_illumina_hiseq_x_ten',
    'instrument_illumina_iseq_100',
    'instrument_illumina_miniseq',
    'instrument_illumina_miseq',
    'instrument_illumina_novaseq_6000',
    'instrument_illumina_novaseq_x_plus',
    'instrument_ion_genesudio_s5',
    'instrument_ion_genesudio_s5_plus',
    'instrument_ion_genesudio_s5_prime',
    'instrument_ion_s5',
    'instrument_ion_s5_xl',
    'instrument_ion_torrent_genexus',
    'instrument_ion_torrent_pgm',
    'instrument_ion_torrent_proton',
    'instrument_ion_torrent_s5',
    'instrument_ion_torrent_s5_xl',
    'instrument_mgiseq_2000rs',
    'instrument_minion',
    'instrument_nextseq_1000',
    'instrument_nextseq_2000',
    'instrument_nextseq_500',
    'instrument_nextseq_550',
    'instrument_onso',
    'instrument_pacbio_rs',
    'instrument_pacbio_rs_ii',
    'instrument_promethion',
    'instrument_revio',
    'instrument_sentosa_sq301',
    'instrument_sequel',
    'instrument_sequel_ii',
    'instrument_sequel_iie',
    'instrument_tapestri',
    'instrument_ug_100',
    'instrument_unspecified',
    'study_type_cancer_genomics',
    'study_type_epigenetics',
    'study_type_exome_sequencing',
    'study_type_metagenomics',
    'study_type_other',
    'study_type_pooled_clone_sequencing',
    'study_type_population_genomics',
    'study_type_synthetic_genomics',
    'study_type_transcriptome_analysis',
    'study_type_transcriptome_sequencing',
    'study_type_whole_genome_sequencing',
]
"""

"""
STRATEGY
    # Genomic
    'wgs': 'Whole Genome Sequencing',
    'wga': 'Whole Genome Amplification',
    'wxs': 'Whole Exome Sequencing',
    'targeted_capture': 'Targeted Capture sequencing',
    'synthetic_long_read': 'Synthetic Long Read sequencing',
    'gbs': 'Genotyping by Sequencin',
    'rad_seq': 'Restriction Site Associated DNA Sequencing',
    'tn_seq': 'Tn sequencing, potentially referring to Transposon Sequencing',
    'clone_end': 'Clone End sequencing',
    # Genomic or Transcriptomic,
    'amplicon': 'Amplicon sequencing',
    'clone': 'Cloning sequencing',
    # Transcriptomic
    'rna_seq': 'RNA sequencing',
    'mrna_seq': 'messenger RNA sequencing',
    'ncrna_seq': 'non-coding RNA sequencing',
    'ribo_seq': 'Ribosome profiling',
    'rip_seq': 'RNA Immunoprecipitation sequencing',
    'mirna_seq': 'microRNA sequencing',
    'ssrna_seq': 'single-stranded RNA sequencing',
    'est': 'Expressed Sequence Tag sequencing',
    'fl_cdna': 'Full-Length complementary DNA sequencing',
    # Epigenomic - Chromatin Accessibility
    'atac_seq': 'Assay for Transposase-Accessible Chromatin sequencing',
    'dnase_hypersensitivity': 'DNase Hypersensitivity sequencing',
    'faire_seq': 'Formaldehyde-Assisted Isolation of Regulatory Elements sequencing',
    # Epigenomic - DNA-Protein Binding
    'chip_seq': 'Chromatin Immunoprecipitation Sequencing',
    'chip': 'Chromatin Immunoprecipitation',
    # Epigenomic - Methylome
    'mre_seq': 'Methylation-Sensitive Restriction Enzyme sequencing',
    'bisulfite_seq': 'Bisulfite sequencing',
    'mbd_seq': 'Methyl-CpG Binding Domain sequencing',
    'medip_seq': 'Methylated DNA Immunoprecipitation sequencing',
    # Epigenomic - Chromatin Interaction and Structural Analysis
    'hi_c': 'Hi-C sequencing for three-dimensional genome structure analysis',
    'chia_pet': 'Chromatin Interaction Analysis by Paired-End Tag Sequencing',
    'tethered_chromatin_conformation_capture': 'Tethered Chromatin Conformation Capture',
    """
